Compute tangents if not provided

jjcasmar opened the pull request:
glTF 2.0 requires we compute tangents if necessary using MikkTSpace. 
We could only generate them if the material for a primitive has a normal map, but since we can change on runtime the material, I have chosen to compute them always, is possible. 

Probably still need some massage for making it faster.

er-vin commented on src/core/gltf2importer/gltf2importer.pri:
```diff
@@ -43,7 +43,8 @@ SOURCES += \
     $$PWD/materialparser.cpp \
-    $$PWD/skinparser.cpp
+    $$PWD/skinparser.cpp \
+    $$PWD/mikktspace/mikktspace.c
```
Could it go in a 3rdparty folder inside of src?

jjcasmar replied:
Im not sure how to do this. Do I have to add a new .pri file?

er-vin replied:
Probably best... but maybe a bit involved for just a file. Or you could use relative paths too. Might have to do something regarding the include path in any cases.

jjcasmar replied:
Done. Thanks Paul.

er-vin commented on src/core/gltf2importer/mikktspace/mikktspace.c:
```diff
@@ -0,0 +1,1843 @@
+/**
+ *  Copyright (C) 2011 by Morten S. Mikkelsen
+ *
+ *  This software is provided 'as-is', without any express or implied
```
I have some concerns about the license here. Since I'm not a lawyer, I've a hard time to figure out if it's really compatible with AGPLv3 in practice. My gut feeling is that it is... but my guts aren't a recognized lawyer.

jcelerier replied:
it's the zlib license so I am pretty sure it is fine - it's already used in various Qt 3rdparty dependencies

jjcasmar replied:
Who should we ask? Looks like is possible, but an opinion from an expert would be nice.

er-vin replied:
Oh right, that's the zlib license. Somehow didn't recognize it. We're fine then.

er-vin commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+}
+
+
+std::array<unsigned int, 3> findVertexIndicesInFace(const Qt3DRender::QAttribute *indexAttr, const int iFace) {
```
Did you run clang-format? This curly brace should be on its own line.

er-vin replied:
Also, dare I ask for unit tests? :-)

jjcasmar replied:
Looks like clang-format was disabled

er-vin commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+std::array<unsigned int, 3> findVertexIndicesInFace(const Qt3DRender::QAttribute *indexAttr, const int iFace) {
+    const auto byteOffset = indexAttr->byteOffset();
+    const auto byteStride = indexAttr->byteStride() == 0 ?
+                            3*vertexBaseTypeSize(indexAttr->vertexBaseType()) :
```
Spaces around *

There are more of those, but I'll ignore style issues in the rest of this review. Really, get the clang-format integration with git commit or your IDE so that it does it for you.

er-vin commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+    };
+
+    interface.m_getPosition = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
```
Indentation looks wrong here

jcelerier replied:
could be the array declaration in the lambda parameters that threw the auto-indenter out ?

er-vin commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+                    float fvPosOut[],
+                    const int iFace,
+                    const int iVertex) {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
```
Lots of duplication around those m_get*, also lots of the work is done more than once. I think we should aim at de-duplicating and reifying the computation in the factored code to make it faster. It's all based on already known addresses and properties of the various attributes at the creation of those lambda, let's lock the lambda behavior at that creation time.

In most of those the only missing information is vertexIndex, all the rest should be already known.

jjcasmar replied:
So Im adding some helper structs that can be used to cache some searchs, have less duplication and make it faster

jcelerier commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        const auto vertexBuffer = userData->positionAttribute->buffer();
+        const auto byteOffset = userData->positionAttribute->byteOffset();
+        const auto byteStride = userData->positionAttribute->byteStride() == 0 ?
+                                3*4 : // Position vertex base type must be float
```
maybe an assert to ensure this ?

jjcasmar replied:
Yes, but not here. Attributes have a specific type required by gltf. We should check there that we are using the correct types.

jjcasmar replied:
I have added a new commit to the PR that checks if the mesh is valid. That way, when we generate the tangents, we have already checked all this.

jcelerier commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        const auto vertexIndex = vertexIndices[iVertex];
+        auto vertexBuffer = userData->tangentAttribute->buffer();
+        const auto byteOffset = 0;
+        const auto byteStride = 4*4;
```
the 4*4 looks fairly magic :p

jjcasmar replied:
You are right. I have rename it

jjcasmar commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+    case Qt3DRender::QAttribute::UnsignedShort:
+        return 2;
+    case Qt3DRender::QAttribute::UnsignedInt:
+        return 4;
```
glTF requires this attribute types to ocupy this space. As you said, there is a problem if in a particular system UnsignedInt is an unsigned int which occupies 2bytes for example. Should we use uint_32t and this kind of types?

jjcasmar replied:
It looks like Qt doesnt work on machines that have strange sizes for some data types. I have found at least a check for float and int in qt codebase

jjcasmar commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        const auto *typedIndices = reinterpret_cast<const unsigned char *>(iFaceHead.data());
+        vertexIndices[0] = typedIndices[0];
+        vertexIndices[1] = typedIndices[1];
+        vertexIndices[2] = typedIndices[2];
```
The face exists because I have create the buffers. If something is wrong, should be signaled before.

jjcasmar commented:
I have redo much of the work in this PR. First of all, I have added some checks to verify that the attributes that we generate are valid according to glTF2.0. This is necessary as Im assuming that the attributes have specific sizes and data types. 

I have precompute most of the data that is needed in the generation of tangents and added a helper structure to avoid finding several times the vertex indices for a particular face.

mkrus commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +210,284 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+    mutable std::array<unsigned int, 3> verticesForLastFace;
+};
+
+struct MikkTSpaceUserData {
```
clang reports warnings about adding padding to keep memory alignment to members of this struct. Moving `primitiveType` to before `vertexIndicesFinder` fixes those.

This (along with a lot of the pre-existing code in this file) produces warnings of signed/unsigned conversion in places that mix int and uint. `nFaces` and `nVertices` should be uint here I think

mkrus commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +210,284 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+    {
+    }
+
+    const std::array<unsigned int, 3> &operator()(int iFace) const
```
#include <array> is missing for me...

jjcasmar commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+    switch (indexAttr->vertexBaseType()) {
+    case Qt3DRender::QAttribute::UnsignedByte: {
+        const auto *typedIndices = reinterpret_cast<const unsigned char *>(iFaceHead.data());
+        vertexIndices[0] = typedIndices[0];
```
I dont understand the second comment

jjcasmar commented:
I have moved the validation and tangent creation code to meshparser_utils.cpp

lemirep commented:
Could you had a reference to the 3rd party in the Readme please

jjcasmar commented on src/core/animationplayer.cpp:
```diff
@@ -260,6 +262,9 @@ AnimationPlayer::AnimationPlayer(Qt3DCore::QNode *parent)
     });
 
     connect(m_animator, &QClipAnimator::runningChanged, this, &AnimationPlayer::runningChanged);
+    connect(m_animator, &QClipAnimator::runningChanged, []() {
```
In fact, this shouldn't be here at all. Was some debugging code for other patch

jjcasmar commented on src/core/gltf2importer/meshparser_utils.cpp:
```diff
@@ -0,0 +1,381 @@
+             Qt3DRender::QAttribute::UnsignedInt };
+}
+
+std::vector<uint> validVertexSizesForAttribute(const QString &attributeName)
```
I would prefer to leave it like this. I would need to take into account that -1 in the check after this, which makes it less clear. In most cases this std::vector is not going to perform any allocation anyway, no?

lemirep replied:
Well std::vector is always going to perform an allocation. What it does though is probably preallocate a given capacity so that it doesn't have to reallocate every time you insert a new elements. So I guess in the end the question is more about allocation time vs if branch. Note that QVarLengthArray might be interesting in this case as you can specify the maximum amount it should contain but fill in less elements than that if you want to which I guess would allow to not do allocations and not have to perform checks.

jjcasmar replied:
You are right about std::vector. I thought it has some small buffer optimization or something like that in common implementations as std::string has.

jjcasmar replied:
I have moved to QVarLengthArray

jjcasmar commented on src/core/gltf2importer/mikktspace/mikktspace.h:
```diff
@@ -0,0 +1,145 @@
+#define __MIKKTSPACE_H__
+
+
+#ifdef __cplusplus
```
You are right, we dont need to export it to C. Still, if we modify the file, we need to document it according to the license. I just copy paste this files from the Blender repo.

lemirep replied:
Alright, let's keep it as is then

jjcasmar commented on src/core/gltf2importer/meshparser_utils.cpp:
```diff
@@ -0,0 +1,381 @@
+
+namespace {
+
+std::vector<Qt3DRender::QAttribute::VertexBaseType> validVertexBaseTypesForAttribute(const QString &attributeName)
```
Same comment as in the std::array<2> above. Im open to discuss it anyway

jjcasmar commented on src/core/gltf2importer/meshparser_utils.cpp:
```diff
@@ -0,0 +1,381 @@
+
+    Attribute indexAttribute;
+    Qt3DRender::QAttribute::VertexBaseType vertexBaseType;
+    mutable int lastFace;
```
Im using GSL recommenation here. We are using a very similar approach in the RangeFinder in Qt3D animation aspect. Maybe I could make them private? 

The function must be const, as Im getting a const ptr to the UserData, which gives me a const view of this object. 

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-fct

lemirep replied:
Ok let's keep it like that then

jjcasmar commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -182,8 +185,34 @@ bool MeshParser::parse(const QJsonArray &meshArray, GLTF2Context *context)
 
+            if (!Kuesa::GLTF2Import::MeshParserUtils::geometryIsGLTF2Valid(geometry)) {
+                qCWarning(kuesa) << QLatin1Literal("Geometry doesn't meet glTF 2.0 requirements");
+                delete geometry;
```
Changed to use a std::unique_ptr

jjcasmar commented on src/core/gltf2importer/meshparser_utils.cpp:
```diff
@@ -0,0 +1,381 @@
+#include "meshparser_utils_p.h"
```
Fixed

jjcasmar commented:
rebased
